File: misc/mkspiral/mkspiral.c
/*****************************
 **
 **  mkspiral.c: A program
 **  to generate training
 **  data for a network with
 **  2 inputs and 1 output.
 **
 **  Any questions or comment
 **  on this task contact:
 **    Alexis P. Wieland
 **    MITRE Corporation
 **    (703) 883-7476
 **    wieland@mitre.ARPA
*****************************/

#include <stdio.h>
#include <math.h>

main()
{
    int i;
    double x, y, angle, radius;

    /* write spiral of data */
    for (i = 0; i <= 96; i++) {
        angle = i * M_PI / 16.0;
        radius = 6.5 * (104 - i) / 104.0;

        x = radius * sin(angle);
        y = radius * cos(angle);

        printf("%8.5f %8.5f %3.1f\n", x, y, 1.0);
        printf("%8.5f %8.5f %3.1f\n", -x, -y, 0.0);
    }
}

File: misc/mkspiral/mkspiral.py
# mkspiral.py

import math
import numpy as np
import matplotlib.pyplot as plt

def mkspiral(a=96, b=16, c=6.5, d=104):
    '''
        Generates training data for a network with 2 inputs and 1 output.

        Adapted from mkspiral.c by Alexis P. Wieland of the MITRE Corporation.
    '''
    ds = []

    for i in range(a + 1):
        angle = i * math.pi / b
        radius = c * (d - i) / d

        x = radius * math.sin(angle)
        y = radius * math.cos(angle)

        ds.append([x, y, 1])
        ds.append([-x, -y, 0])
    
    return np.array(ds)

ds = mkspiral()

X = ds[:,:-1]
y = ds[:,-1]

scatter = plt.scatter(X[:,0], X[:,1], c=y, cmap="grey")

plt.xlabel("$x$")
plt.ylabel("$y$")
plt.title("Spiral Benchmark Test")
plt.legend(scatter.legend_elements()[0], [0, 1])

ax = plt.gca()
ax.set_facecolor("grey")
ax.set_aspect(aspect="equal")

plt.show()